ghcide/src/Development/IDE/Core/ProgressReporting.hs

```diff
@@ -33,119 +34,170 @@ import           Language.LSP.Server            (ProgressAmount (..),
                                                  withProgress)
 import qualified Language.LSP.Server            as LSP
 import qualified StmContainers.Map              as STM
-import           UnliftIO                       (Async, async, cancel)
+import           UnliftIO                       (Async, MonadUnliftIO, async,
+                                                 bracket, cancel)
 
 data ProgressEvent
-    = KickStarted
-    | KickCompleted
+  = ProgressNewStarted
+  | ProgressCompleted
+  | ProgressStarted
 
-data ProgressReporting  = ProgressReporting
-  { progressUpdate :: ProgressEvent -> IO ()
-  , inProgress     :: forall a. NormalizedFilePath -> Action a -> Action a
-  , progressStop   :: IO ()
+data ProgressReporting m = ProgressReporting
+  { progressUpdate :: ProgressEvent -> m (),
+    inProgress     :: forall a. NormalizedFilePath -> m a -> m a,
+    progressStop   :: IO ()
```

Comment:
why is this not in m?

Reply:
`shakeShut :: IdeState -> IO ()` use `progressStop`.